api/discovery/client.go

```diff
@@ -146,3 +147,30 @@ func (s *ClientObserverSet) notifyOne(
 
 	g.Wait()
 }
+
+// ClientExecutor is a ClientObserver that executes a function in a new
+// goroutine whenever a client connects.
+type ClientExecutor struct {
+	executor
+
+	// Func is the function to execute when a client connects.
+	// The context is canceled when the target becomes unavailable.
+	Func func(context.Context, *Client)
```

Comment:
I have noticed that this function does not return an error. Is `ClientExecutor` not interested what this function returns? On the other hand I understand that this is just a connection hook. Same is related to `TargetExecutor.Func()` method.

Reply:
Yep, I debated this for a little while, but ultimately there's nothing useful I can do with an error in this type, so it's best if we force the user to handle it inside their function by whatever means is appropriate.